File: legacy/policy-service/src/main/java/pl/altkom/asc/lab/micronaut/poc/policy/infrastructure/config/PostgresDatasourceConfigurationListener.java
package pl.altkom.asc.lab.micronaut.poc.policy.infrastructure.config;

import io.micronaut.configuration.jdbc.hikari.DatasourceConfiguration;
import io.micronaut.context.annotation.Requires;
import io.micronaut.context.env.Environment;
import io.micronaut.context.event.BeanCreatedEvent;
import io.micronaut.context.event.BeanCreatedEventListener;

import javax.inject.Singleton;
import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;

@Singleton
@Requires(notEnv = Environment.TEST)
public class PostgresDatasourceConfigurationListener
        implements BeanCreatedEventListener<DatasourceConfiguration> {

    private final PostgresDatasourceProperties postgres;

    public PostgresDatasourceConfigurationListener(PostgresDatasourceProperties postgres) {
        this.postgres = postgres;
    }

    @Override
    public DatasourceConfiguration onCreated(BeanCreatedEvent<DatasourceConfiguration> event) {
        DatasourceConfiguration cfg = event.getBean();

        if ("default".equals(cfg.getName())) {
            cfg.setDriverClassName("org.postgresql.Driver");
            cfg.setUsername(postgres.getUsername());
            cfg.setPassword(postgres.getPassword());
            cfg.setUrl(buildJdbcUrl());
        }

        return cfg;
    }

    private String buildJdbcUrl() {
        String sslMode = postgres.isSsl() ? "require" : "disable";

        return "jdbc:postgresql://" + postgres.getHost() + ":" + postgres.getPort()
                + "/" + postgres.getDatabase()
                + "?sslmode=" + sslMode
                + "&user=" + encode(postgres.getUsername())
                + "&password=" + encode(postgres.getPassword());
    }

    private String encode(String value) {
        return URLEncoder.encode(value, StandardCharsets.UTF_8);
    }
}

File: legacy/documents-service/src/main/kotlin/pl/altkom/asc/lab/micronaut/poc/documents/infrastructure/config/PostgresDatasourceConfigurationListener.kt
package pl.altkom.asc.lab.micronaut.poc.documents.infrastructure.config

import io.micronaut.configuration.jdbc.hikari.DatasourceConfiguration
import io.micronaut.context.event.BeanCreatedEvent
import io.micronaut.context.event.BeanCreatedEventListener
import java.net.URLEncoder
import java.nio.charset.StandardCharsets
import javax.inject.Singleton


@Singleton
class PostgresDatasourceConfigurationListener(
    private val postgres: PostgresDatasourceProperties
) : BeanCreatedEventListener<DatasourceConfiguration> {

    override fun onCreated(event: BeanCreatedEvent<DatasourceConfiguration>): DatasourceConfiguration {
        val cfg = event.bean

        if (cfg.name == "default") {
            cfg.driverClassName = "org.postgresql.Driver"
            cfg.username = postgres.username
            cfg.password = postgres.password
            cfg.url = buildJdbcUrl()
        }
       return cfg
    }

    private fun buildJdbcUrl(): String {
        val user = encode(postgres.username)
        val password = encode(postgres.password)
        val sslMode = if (postgres.ssl) "require" else "disable"

        return "jdbc:postgresql://${postgres.host}:${postgres.port}/${postgres.database}" +
                "?sslmode=$sslMode&user=$user&password=$password"
    }

    private fun encode(value: String): String =
        URLEncoder.encode(value, StandardCharsets.UTF_8.toString())
}
